core/src/test/scala/unit/kafka/admin/DeleteTopicTest.scala

```diff
@@ -187,164 +197,233 @@ class DeleteTopicTest extends QuorumTestHarness {
+      TestUtils.resource(Admin.create(props)) { adminClient =>
+        try {
+          adminClient.createPartitions(Map(topic -> NewPartitions.increaseTo(2)).asJava).all().get()
+        } catch {
+          case _: ExecutionException =>
+        }
       }
+      // trigger a controller switch now
+      val previousControllerId = controllerId
+
+      controller.shutdown()
+
+      ensureControllerExists()
+      // wait until a new controller to show up
+      TestUtils.waitUntilTrue(() => {
+        val (newController, newControllerId) = getController()
+        newControllerId != previousControllerId
+      }, "The new controller should not have the failed controller id")
+
+      // bring back the failed brokers
+      follower.startup()
+      controller.startup()
+      TestUtils.verifyTopicDeletion(zkClient, topic, 2, partitionHostingServers)
     }
-    // trigger a controller switch now
-    val previousControllerId = controllerId
-
-    controller.shutdown()
-
-    ensureControllerExists()
-    // wait until a new controller to show up
-    TestUtils.waitUntilTrue(() => {
-      val (newController, newControllerId) = getController()
-      newControllerId != previousControllerId
-    }, "The new controller should not have the failed controller id")
-
-    // bring back the failed brokers
-    follower.startup()
-    controller.startup()
-    TestUtils.verifyTopicDeletion(zkClient, topic, 2, servers)
   }
 
-
-  @Test
-  def testDeleteTopicDuringAddPartition(): Unit = {
-    val topic = "test"
-    servers = createTestTopicAndCluster(topic)
-    val leaderIdOpt = zkClient.getLeaderForPartition(new TopicPartition(topic, 0))
-    assertTrue(leaderIdOpt.isDefined, "Leader should exist for partition [test,0]")
-    val follower = servers.filter(_.config.brokerId != leaderIdOpt.get).last
+  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumName)
+  @ValueSource(strings = Array("zk", "kraft"))
+  def testDeleteTopicDuringAddPartition(quorum: String): Unit = {
+    brokers = createTestTopicAndCluster(topic)
+    val leaderIdOpt = TestUtils.waitUntilLeaderIsKnown(brokers, new TopicPartition(topic, 0))
+    val follower = brokers.filter(_.config.brokerId != leaderIdOpt).last
     val newPartition = new TopicPartition(topic, 1)
-    // capture the brokers before we shutdown so that we don't fail validation in `addPartitions`
-    val brokers = adminZkClient.getBrokerMetadatas()
-    follower.shutdown()
-    // wait until the broker has been removed from ZK to reduce non-determinism
-    TestUtils.waitUntilTrue(() => zkClient.getBroker(follower.config.brokerId).isEmpty,
-      s"Follower ${follower.config.brokerId} was not removed from ZK")
-    // add partitions to topic
-    adminZkClient.addPartitions(topic, expectedReplicaFullAssignment, brokers, 2,
-      Some(Map(1 -> Seq(0, 1, 2), 2 -> Seq(0, 1, 2))))
+
+    if (isKRaftTest()) {
```

Comment:
We have placed the code from both modes in two different branches, which is no different from the two methods. Can we try to ensure that the same code is used in both modes as much as possible. for example, the 2 branches both have a `follower.shutdown()` and both call `increasePartitions`, we can move it out of it/else and move `adminZkClient.addPartitions` into `TestUtils.increasePartitions`, WDYT.

Reply:
The type of `follower` object is different, depending on which mode. In KRaft it is KafkaBroker and in ZK it is KafkaServer, even though they both support shutdown() method, I'm not sure we can define this object with different types. I did try this, but ended up with many if statements and it didn't quite work. So I placed them into only two branches since the only common code was increasePartitions(). I'm happy to move increasePartitions() out of the if else.